### PR TITLE
Basic User management

### PR DESCRIPTION
**Pull request by CDimonaco**
# Description

This pr adds a basic user management functionality

- Enrich the User schema with new fields
- Enforce security best practices on user schema in terms of passwords hashing
- Create the phoenix context for crud operations on user schema
- Implements soft delete of users
- Implement locking of users
- Instrument pow framework to deal with soft deleted users and locked users on login
- Updates the seed and release task to generate a proper admin user with the new fields and the new password with the enforced Argon2 algo
- Creates all the controllers for crudding user resource, after each update the involved user should be broadcasted on the user dedicated channel
- Implements user socket authentication
- Create dedicated socket channels per user to broadcast user specific event

## How was this tested?

Automated tests

### CODE REVIEW

File: config/config.exs [hunk 1]
```diff
@@ -168,8 +169,8 @@ config :trento, :jwt_authentication,
   issuer: "https://github.com/trento-project/web",
   app_audience: "trento_app",
   api_key_audience: "trento_api_key",
-  # Seconds, 10 minutes
-  access_token_expiration: 600,
+  # Seconds, 3 minutes
+  access_token_expiration: 180,
```
Comment (EMaksy) [inline config/config.exs:173]: Just curious why make it shorter ?

File: lib/trento/users.ex [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    end
+  end
+
+  def create_user(attrs \\ %{}) do
+    %User{}
+    |> User.changeset(attrs)
+    |> Repo.insert()
+  end
+
+  def update_user(%User{id: 1}, _), do: {:error, :operation_not_permitted}
```
Comment (EMaksy) [inline lib/trento/users.ex:50]: Maybe a comment that this is the admin user which must not be deleted

File: priv/repo/migrations/20240416091937_add_email_and_fullname_to_user.exs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+defmodule Trento.Repo.Migrations.AddEmailAndFullnameToUser do
```
Comment (EMaksy) [inline priv/repo/migrations/20240416091937_add_email_and_fullname_to_user.exs:1]: Should the name of the alias not be ```defmodule ... AddEmailAndFullNameToUser do .....``` ?

File: config/config.exs (hunk 1 above)
Comment (arbulu89) [reply config/config.exs:173]: In the new context where we have users, and we can remove/lock them, this deleting/locking takes effect immediately in the database, but as they authentication jwt token is stateless, you cannot change it.
So when you lock a user, the obtained access key would still be valid.
Reducing this time limits this "grace period" of 10 minutes to 3, so "not wanted users" validity time is reduced.
It is just a safety measure.

File: lib/trento/release.ex [hunk 4]
```diff
@@ -71,15 +70,19 @@ defmodule Trento.Release do
 
     admin_user = System.get_env("ADMIN_USER", "admin")
     admin_password = System.get_env("ADMIN_PASSWORD", "adminpassword")
+    admin_email = System.get_env("ADMIN_EMAIL", "admin@trento.suse.com")
```
Comment (arbulu89) [inline lib/trento/release.ex:73]: Let's discuss this default value with @abravosuse

File: lib/trento/users.ex [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    user: Trento.Users.User
+
+  import Ecto.Query, warn: false
+  alias Trento.Repo
+
+  alias Trento.Users.User
+
+  @impl true
+  @doc """
+    get_by function overrides the one defined in Pow.Ecto.Context,
```
Comment (arbulu89) [inline lib/trento/users.ex:17]: Let's remove this trailling tab

File: lib/trento/users.ex [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    we retrieve the user by username as traditional Pow flow but we also exclude
+    deleted and locked users
+  """
+  def get_by(clauses) do
+    username = clauses[:username]
+
+    User
+    |> where([u], is_nil(u.deleted_at) and is_nil(u.locked_at) and u.username == ^username)
```
Comment (arbulu89) [inline lib/trento/users.ex:25]: Are we sure we want to filter out locked users?
We don't filter out them in the list for example.

File: lib/trento/users.ex [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    |> Repo.one()
+  end
+
+  def list_users do
+    User
+    |> where([u], is_nil(u.deleted_at))
+    |> Repo.all()
+  end
+
+  def get_user(id) do
```
Comment (arbulu89) [inline lib/trento/users.ex:35]: I guess this `get_user` is a function out of the standard `crud`, right?
To be used outside the controller functions.
Or maybe we don't need the `get_by` above.
Just asking hehe

File: lib/trento/users.ex [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+  end
+
+  def update_user(%User{id: 1}, _), do: {:error, :operation_not_permitted}
+
+  def update_user(%User{locked_at: nil} = user, %{enabled: false} = attrs) do
+    do_update(user, Map.put(attrs, :locked_at, DateTime.utc_now()))
+  end
+
+  def update_user(%User{locked_at: locked_at} = user, %{enabled: false} = attrs)
+      when not is_nil(locked_at) do
```
Comment (arbulu89) [inline lib/trento/users.ex:57]: You don't need this `is_nil`. With the pattern matching on top of this, you are assured that it never will be

File: lib/trento_web/channels/user_channel.ex [hunk 9]
```diff
@@ -0,0 +1,34 @@
+defmodule TrentoWeb.UserChannel do
+  @moduledoc """
+  User channel, each use is subscribed to his channel,
```
Comment (arbulu89) [inline lib/trento_web/channels/user_channel.ex:3]: typo `use` -> `user`

File: lib/trento_web/channels/user_channel.ex [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+  to receive personal broadcasts
+
+  Users can't join other users channels.
```
Comment (arbulu89) [inline lib/trento_web/channels/user_channel.ex:6]: Better `Users can't join other users channel`.
And I don't really understand well the next phrase `This has a prerequisite...`

File: lib/trento_web/openapi/v1/schema/user.ex [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    OpenApiSpex.schema(%{
+      title: "UserCreationRequest",
+      description: "Request body to create a user",
+      type: :object,
+      additionalProperties: false,
+      properties: %{
+        fullname: %Schema{type: :string, description: "User full name", nullable: false},
+        email: %Schema{type: :string, description: "User email", nullable: false, format: :email},
+        username: %Schema{type: :string, description: "User username", nullable: false},
+        enabled: %Schema{
```
Comment (arbulu89) [inline lib/trento_web/openapi/v1/schema/user.ex:19]: is `enabled` here the opposite of `locked`?
It is maybe a bit strange to use `enabled `term in the controller and `locked` in the "backend/schemas".
But i undestand the terms anyway

File: lib/trento_web/openapi/v1/schema/user.ex [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+          nullable: false
+        },
+        updated_at: %OpenApiSpex.Schema{
+          type: :string,
+          format: :"date-time",
+          description: "Date of user last update",
+          nullable: true
+        }
+      },
+      required: [:id, :fullname, :email, :created_at]
```
Comment (arbulu89) [inline lib/trento_web/openapi/v1/schema/user.ex:104]: the `username` is not required?

File: lib/trento_web/plugs/app_jwt_auth_plug.ex [hunk 13]
```diff
@@ -101,4 +99,25 @@ defmodule TrentoWeb.Plugs.AppJWTAuthPlug do
   @spec validate_refresh_token(binary()) :: {atom(), any()}
   defp validate_refresh_token(jwt_token),
     do: RefreshToken.verify_and_validate(jwt_token)
+
+  def user_allowed_to_renew?(%User{deleted_at: deleted_at, locked_at: locked_at})
```
Comment (arbulu89) [inline lib/trento_web/plugs/app_jwt_auth_plug.ex:103]: This function should come after `attach_refresh_token_to_conn`?
And are we sure it should be a public function?

File: lib/trento_web/views/session_view.ex [hunk 14]
```diff
@@ -11,7 +11,8 @@ defmodule TrentoWeb.SessionView do
 
   def render("me.json", %{user: user}) do
     %{
-      username: user.username
+      username: user.username,
```
Comment (arbulu89) [inline lib/trento_web/views/session_view.ex:14]: I would obtain the `username` and `id` in the function signature. It looks prettier

File: lib/trento_web/views/v1/user_view.ex [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+    render_many(users, __MODULE__, "user.json")
+  end
+
+  def render("show.json", %{user: user}) do
+    render_one(user, __MODULE__, "user.json")
+  end
+
+  def render("user.json", %{user: user}) do
+    %{
+      id: user.id,
```
Comment (arbulu89) [inline lib/trento_web/views/v1/user_view.ex:14]: The same, maybe better destructing the fields in the signature, it looks more beatiful

File: priv/repo/seeds.exs [hunk 16]
```diff
@@ -14,7 +14,10 @@
 |> Trento.Users.User.changeset(%{
   username: "admin",
   password: "adminpassword",
-  confirm_password: "adminpassword"
+  confirm_password: "adminpassword",
+  fullname: "Trento Admin",
+  email: "admin@trento.com",
```
Comment (arbulu89) [inline priv/repo/seeds.exs:19]: Here we are using a different email than the release task hehe
In the release we have `admin@trento.suse.com`

File: test/trento/users_test.exs [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+      refute username == original_username
+    end
+  end
+
+  # We can't use factory since we have to deal with pow logic
+  defp create_user do
+    password = "themightypassword8897"
+
+    {:ok, user} =
+      Users.create_user(%{
```
Comment (arbulu89) [inline test/trento/users_test.exs:200]: Cannot we use a factory here?
You are implementing one at least hehe
I think it would be appropriate

File: test/trento/users_test.exs [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+        fullname: Faker.Pokemon.name(),
+        password: password,
+        password_confirmation: password,
+        username: Faker.Pokemon.name()
+      })
+
+    user
+  end
+
+  defp create_deleted_user do
```
Comment (arbulu89) [inline test/trento/users_test.exs:211]: you could do this with the factory as well, liks `set_deleted` or other function if you want

**Review by arbulu89 — COMMENTED** [review]
@CDimonaco Ultra work! kudos
I left a lot of comments/questions.
I would like to clarify them before approving, but so far so good.

My unique "red flag" (maybe I'm missing something):
- I think we should create a proper `user` factory to test

File: lib/trento_web/controllers/v1/user_controller.ex [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+    tags: ["User Management"],
+    parameters: [
+      id: [
+        in: :path,
+        required: true,
+        type: %OpenApiSpex.Schema{type: :integer}
+      ]
+    ],
+    responses: [
+      not_found: Schema.NotFound.response(),
```
Comment (rtorrero) [inline lib/trento_web/controllers/v1/user_controller.ex:108]: Maybe I'm missing something but shouldn't we mention here a successful response also?

File: lib/trento_web/channels/user_channel.ex [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+  require Logger
+  use TrentoWeb, :channel
+
+  @impl true
+  def join(
+        "users:" <> user_id,
+        _payload,
+        %{assigns: %{current_user_id: current_user_id}} = socket
+      ) do
+    user_id = String.to_integer(user_id)
```
Comment (EMaksy) [inline lib/trento_web/channels/user_channel.ex:18]: General question why are we using integer ID and not UUID ?
I guess to keep it simple ?

**Review by rtorrero — APPROVED** [review]
Some minor things but LGTM! Thanks @CDimonaco

File: lib/trento/users.ex (hunk 6 above)
Comment (CDimonaco) [reply lib/trento/users.ex:25]: Locked out users could not login and the get by callback is just an override of pow context to filters users on login

File: lib/trento/users.ex (hunk 7 above)
Comment (CDimonaco) [reply lib/trento/users.ex:35]: The function is an override of the Pow Context to customize the user retrieval in login

File: lib/trento_web/openapi/v1/schema/user.ex (hunk 11 above)
Comment (CDimonaco) [reply lib/trento_web/openapi/v1/schema/user.ex:19]: Yeah it's true, on the frontend/api we use enabled so I stick with it but we can change

File: priv/repo/migrations/20240416091937_add_email_and_fullname_to_user.exs (hunk 3 above)
Comment (CDimonaco) [reply priv/repo/migrations/20240416091937_add_email_and_fullname_to_user.exs:1]: Phoenix generator generated the migration, so we keep like that :eyes:

File: lib/trento_web/controllers/v1/user_controller.ex [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+    ]
+
+  def show(conn, %{id: id}) do
+    with {:ok, user} <- Users.get_user(id) do
+      render(conn, "show.json", user: user)
+    end
+  end
+
+  operation :update,
+    summary: "Create an existing user",
```
Comment (EMaksy) [inline lib/trento_web/controllers/v1/user_controller.ex:73]: Changing  "Create an existing user"  to "Update an existing user" would not harm ?

File: lib/trento_web/openapi/v1/schema/user.ex [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+
+    # see: https://github.com/open-api-spex/open_api_spex/issues/87
+    # This is an alternative way of defining schemas without having to deal to default values
+    # when the struct is converted to plain parameters in controller
+    # by default all the struct values have nil values, so we can't distinguish between a parameter not passed
+    # or a parameter passed explicitly nil.
+    # this is an update request and we have to distinguish in PATCH verb values passed or not passed.
+    # Having a value not passed as nil by default is not practical nor correct.
+    # this means that in the controller the body params are passed as map, without converting from the struct
+    # and we have everything casted and validated BUT without the hassle of cannot distinguish between passed
```
Comment (EMaksy) [inline lib/trento_web/openapi/v1/schema/user.ex:68]: I guess a typo in the tense:

and we have everything ~~casted~~ --> and we have everything **cast**

File: lib/trento/users.ex (hunk 2 above)
Comment (CDimonaco) [reply lib/trento/users.ex:50]: it's not possible to update the admin user, we have environment variables to control the attributes, it will overwrite things in the release task

File: lib/trento_web/controllers/v1/user_controller.ex (hunk 19 above)
Comment (CDimonaco) [reply lib/trento_web/controllers/v1/user_controller.ex:108]: Yeah, but in restful approach it's no content the response on delete, so we don't have any response because it's no content

File: lib/trento/users/user.ex [hunk 23]
```diff
@@ -2,22 +2,112 @@ defmodule Trento.Users.User do
...
-    user_id_field: :username
+    user_id_field: :username,
+    password_hash_methods: {&Argon2.hash_pwd_salt/1, &Argon2.verify_pass/2}
 
   use Pow.Extension.Ecto.Schema,
     extensions: [PowPersistentSession]
 
+  alias EctoCommons.EmailValidator
+
+  @sequences ["01234567890", "abcdefghijklmnopqrstuvwxyz"]
```
[the remark this answers is not in the document]
Comment (CDimonaco) [reply lib/trento/users/user.ex:16]: I followed the standards reccomended by plug based on nist reccomendation, I think we could stick with that validation for now

File: priv/repo/seeds.exs (hunk 16 above)
Comment (CDimonaco) [reply priv/repo/seeds.exs]: done

File: lib/trento_web/views/v1/user_view.ex (hunk 15 above)
Comment (CDimonaco) [reply lib/trento_web/views/v1/user_view.ex]: done

File: lib/trento_web/views/session_view.ex (hunk 14 above)
Comment (CDimonaco) [reply lib/trento_web/views/session_view.ex]: done

File: lib/trento_web/plugs/app_jwt_auth_plug.ex (hunk 13 above)
Comment (CDimonaco) [reply lib/trento_web/plugs/app_jwt_auth_plug.ex]: You're right, moved function and made it private

File: lib/trento_web/openapi/v1/schema/user.ex (hunk 12 above)
Comment (CDimonaco) [reply lib/trento_web/openapi/v1/schema/user.ex]: You're right, added

File: lib/trento_web/channels/user_channel.ex (hunk 10 above)
Comment (CDimonaco) [reply lib/trento_web/channels/user_channel.ex]: adjusted!

File: lib/trento_web/channels/user_channel.ex (hunk 9 above)
Comment (CDimonaco) [reply lib/trento_web/channels/user_channel.ex]: done

File: lib/trento/users.ex (hunk 5 above)
Comment (CDimonaco) [reply lib/trento/users.ex]: done

File: lib/trento_web/openapi/v1/schema/user.ex (hunk 22 above)
Comment (CDimonaco) [reply lib/trento_web/openapi/v1/schema/user.ex:68]: done

File: lib/trento/users.ex (hunk 8 above)
Comment (CDimonaco) [reply lib/trento/users.ex:57]: We need this clause to match the case, when the user is already locked, but from the api comes to "relock" again, to not update the locked_at timestamp, it's an edge case, if I remove that clause the tests does not pass for that edge case

File: test/trento/users_test.exs (hunk 18 above)
Comment (CDimonaco) [reply test/trento/users_test.exs:196]: added factory!

File: test/trento/users_test.exs (hunk 17 above)
Comment (CDimonaco) [reply test/trento/users_test.exs]: added factory!

**Comment by CDimonaco** [comment]
@arbulu89 Addressed feedbacks and factory added!

File: test/trento/users_test.exs [hunk 24]
```diff
@@ -0,0 +1,199 @@
...
+               Users.create_user(%{
+                 username: "username",
+                 password: "some password",
+                 email: "test@trento.com",
+                 fullname: "some fullname",
+                 confirm_password: "some password"
+               })
+
+      assert user.password == nil
+      assert user.fullname == "some fullname"
```
Comment (EMaksy) [inline test/trento/users_test.exs:40]: I guess we also could use faker/factory here to generate random strings

File: lib/trento/users.ex (hunk 8 above)
Comment (arbulu89) [reply lib/trento/users.ex:57]: Ok ok I see. I was not thinking about the lock a locked user hehe

File: test/support/factory.ex [hunk 25]
```diff
@@ -883,4 +885,18 @@ defmodule Trento.Factory do
       to_package_id: "#{RandomElixir.random_between(0, 1000)}"
     }
   end
+
+  def user_factory do
+    password = Faker.Pokemon.name()
+
+    %User{
+      email: Faker.Internet.email(),
```
Comment (arbulu89) [inline test/support/factory.ex:893]: I hope that the uniqueness of username and email doesn't raise any issue here. Hopefully the faker dictionaries have enough emails hehe
Otherwise we can add `sequence` in the future

File: test/trento/users_test.exs [hunk 26]
```diff
@@ -0,0 +1,199 @@
...
+  use Trento.DataCase
+
+  alias Trento.Users
+  alias Trento.Users.User
+  import Trento.Factory
+
+  describe "users" do
+    test "list_users returns all users except the deleted ones" do
+      %{id: user_id} = insert(:user)
+      create_deleted_user()
```
Comment (arbulu89) [inline test/trento/users_test.exs:11]: I don't think we need a function for this. Maybe better to put `insert(:user, deleted_at: DateTime.utc_now())` in the code itself. I find it more declarative

**Review by arbulu89 — APPROVED** [review]
Really great man!

File: test/trento/users_test.exs (hunk 26 above)
Comment (CDimonaco) [reply test/trento/users_test.exs]: done!

File: lib/trento_web/controllers/v1/user_controller.ex (hunk 21 above)
Comment (CDimonaco) [reply lib/trento_web/controllers/v1/user_controller.ex]: done! thx!

**Review by EMaksy — APPROVED** [review]
Magnifico :+1:
